satellite/repair/checker/rangedloop.go

```diff
@@ -0,0 +1,376 @@
+	return queue.NewInsertBuffer(observer.repairQueue, observer.repairQueueBatchSize)
+}
+
+// CompareStats compares total stats with given data, used only for tests.
+func (observer *RangedLoopObserver) CompareStats(objChecked int64, remoteSegmentsChecked int64, remoteSegmentsNeedingRepair int64,
+	newRemoteSegmentsNeedingRepair int64, remoteSegmentsLost int64, remoteSegmentsFailedToCheck int64, objLost []uuid.UUID) error {
+	stats := observer.TotalStats
+	if stats.objectsChecked != objChecked {
+		return errs.New("objects checked amount is different")
+	}
+	if stats.remoteSegmentsLost != remoteSegmentsLost {
+		return errs.New("remote segments lost amount is different")
+	}
+	if stats.remoteSegmentsChecked != remoteSegmentsChecked {
+		return errs.New("remote segments checked amount is different")
+	}
+	if stats.remoteSegmentsNeedingRepair != remoteSegmentsNeedingRepair {
+		return errs.New("remote segments needing repair amount is different")
+	}
+	if stats.newRemoteSegmentsNeedingRepair != newRemoteSegmentsNeedingRepair {
+		return errs.New("new remote segments needing repair amount is different")
+	}
+	if stats.remoteSegmentsFailedToCheck != remoteSegmentsFailedToCheck {
+		return errs.New("new remote segments needing repair amount is different")
+	}
+
+	if !reflect.DeepEqual(stats.objectsLost, objLost) {
+		return errs.New("objects lost amount is different")
+	}
+
+	return nil
+}
+
+// Start starts parallel segments loop.
+func (observer *RangedLoopObserver) Start(ctx context.Context, startTime time.Time) error {
+	return nil
+}
+
+// Fork creates a Partial to process a chunk of all the segments.
+func (observer *RangedLoopObserver) Fork(ctx context.Context) (rangedloop.Partial, error) {
+	return NewRangedLoopCheckerPartial(observer), nil
+}
+
+// Join is called after the chunk for Partial is done.
+// This gives the opportunity to merge the output like in a reduce step.
+func (observer *RangedLoopObserver) Join(ctx context.Context, partial rangedloop.Partial) error {
+	repPartial := partial.(*repairPartial)
+	if err := repPartial.repairQueue.Flush(ctx); err != nil {
+		return Error.Wrap(err)
+	}
+
+	observer.TotalStats.objectsLost = append(observer.TotalStats.objectsLost, repPartial.monStats.objectsLost...)
+	observer.TotalStats.newRemoteSegmentsNeedingRepair += repPartial.monStats.newRemoteSegmentsNeedingRepair
+	observer.TotalStats.remoteSegmentsChecked += repPartial.monStats.remoteSegmentsChecked
+	observer.TotalStats.remoteSegmentsLost += repPartial.monStats.remoteSegmentsLost
+	observer.TotalStats.remoteSegmentsFailedToCheck += repPartial.monStats.remoteSegmentsFailedToCheck
+	observer.TotalStats.remoteSegmentsNeedingRepair += repPartial.monStats.remoteSegmentsNeedingRepair
+	observer.TotalStats.objectsChecked += repPartial.monStats.objectsChecked
+
+	return nil
+}
+
+// Finish is called after all segments are processed by all observers.
+func (observer *RangedLoopObserver) Finish(ctx context.Context) error {
+	startTime := time.Now()
```

Comment:
This is not the start time.

We can add it to the signature of interface `Observer.Finish()` if that's useful. Or read it in RangedLoopObserver.Start() and store it.